### DISCUSSION TITLE
Local Authentication (Biometrics) Portal

### DISCUSSION THREAD

**Discussion by quexten** [flatpak/xdg-desktop-portal#1275] [New Portals]
Some applications - password managers (1Password, Bitwarden), crypto wallets, secure chat applications need to verify that the user present is actually the logged in user upon opening the app. On other operating systems, there are platform APIs for this that authenticate the user using biometrics or the system password, and provide some kind of secret that can be used for en-/de-cryption.

In un-sandboxed applications, the authentication part can somewhat be achieved by installing polkit policies for the application, or by using something like fprintd directly. For sandboxed applications this is not possible.

My specific use-case is implementing biometric auth in Bitwarden in a more sane way that also supports Flatpaks (and Snap), and especially makes it easier in immutable environments (Fedora Silverblue) (https://github.com/bitwarden/clients/pull/4586, https://github.com/quexten/goldwarden/issues/90), but at least 1Password would also benefit here.

## Interactions
There are two basic interactions that are required. These could possibly be just one interface with different options, I'm open to discussion here.

### Local Authentication
This would basically be the same as just calling polkit for authentication. In the implementing frontends, this should show the app (maybe with Icon from the .desktop file?) and a message passed by the caller.

### Local Authentication with Credentials
This should provide a secret, much like the `org.freedesktop.portal.Secret` portal, but locked behind authentication. This would make it possible to have apps locked by biometrics/local authentication and decrypted only upon authentication.

For both of these use-cases, I feel like configurable options for whether to allow password / biometrics makes sense.

**Comment by quexten** [comment]
It seems that on the ui-side, the Gnome team has a relevant mock-up: https://gitlab.gnome.org/Teams/Design/os-mockups/-/issues/232 .

**Comment by Mikenux** [comment]
Is this about unlocking an app so it can be used?

What is expected? Does the application need to acquire the credentials or can we do without them (e.g. set a master password without the application acquiring it?)

> a message passed by the caller.

What type of message are you thinking of?

**Reply by quexten** [reply]
> Is this about unlocking an app so it can be used?

The second part - Local Authentication with Credentials - yes. How it is used is up to the app, but other platform APIs generally provide some kind of secret after authenticating, that *can* be used for encryption. So a password manager could - after enrolling in the settings - encrypt the local vault using the "biometric key" provided by the portal, and upon unlock use this to decrypt.

The former part - Local Authentication (without credentials) is for when an application does not need a secret (maybe it already has them stored in the kernel keyring, maybe it just doesn't need them) and wants to verify that the correct user is present. This could be used for certain - "dangerous"/"protected" actions. For instance, an SSH daemon might protect ssh login / git signing behind this (if configured to do so).

> What is expected? Does the application need to acquire the credentials or can we do without them (e.g. set a master password without the application acquiring it?)

I'm not exactly sure what you mean by this. It's mostly so the application can get a cryptographic key. What it does with this is up to the application, but as mentioned above it *could* be used as an additional "keyslot", next to the regular master password (in the case of a password manager).

> What type of message are you thinking of?

It should be clear who is requesting and why. A generic "please authenticate" message is too broad. Instead it could be something like "Application {application} is requesting permission to do {action}", where the {action} part could be application provided. This is somewhat in line with what the gnome design team did in the linked mock-up.

I can draft up a rough dbus interface for this, but so far I don't have any experience doing this so it might require a few iterations to get right.

**Reply by Mikenux** [reply]
I meant that this doesn't give any personal data to the app (except an app specific secret/key), which I assume is the case given what you said and what is in the mockup.

> Instead it could be something like "Application {application} is requesting permission to do {action}", where the {action} part could be application provided. This is somewhat in line with what the gnome design team did in the linked mock-up.

For portals, giving a reason is generally avoided, unless this action is always the same and one is sure that the app wants to do this action. In mockups, the app wants to verify your identity, and that's it. However, there is no "verify your identity to do {action}". The aim is to prevent the portal from reporting a potentially false reason, even if no personal data is transmitted. The user has to guess based on context, so they can guess if what the app wants makes sense. If you really want this information to be shown in the portal UI, you need to be sure that the app wants to verify your identity to really perform said action.

**Reply by quexten** [reply]
> I meant that this doesn't give any personal data to the app (except an app specific secret/key), which I assume is the case given what you said and what is in the mockup.

Yes, no personal data, only a crypto key / yes/no approval, separated per app/user.

> For portals, giving a reason is generally avoided, unless this action is always the same and one is sure that the app wants to do this action. In mockups, the app wants to verify your identity, and that's it. However, there is no "verify your identity to do {action}". The aim is to prevent the portal from reporting a potentially false reason, even if no personal data is transmitted. The user has to guess based on context, so they can guess if what the app wants makes sense. If you really want this information to be shown in the portal UI, you need to be sure that the app wants to verify your identity to really perform said action.

Yeah, that seems reasonable. On the other hand I don't see any harm if the calling app is also noted (but not provided by the app). What would be such a false reason (under the assumption that the calling app is also shown in the prompt)? On iOS/Android it seems this dialog is also mostly just "verify your identity" (no app/action noted). On MacOS you can specify a reason, f.e from github.com/lox/go-touchid (with the action being "access llamas"): ![Screenshot](https://lachlan.me/s/9TMZWTYGikXoeCHm8RBgi8Bb0o4R1Bz6uI.png)

Either way, for my use-case just verifying identity is enough if providing a reason is deemed a security risk. Though providing an action would be an additional benefit, "use ssh key"/"approve device login"/"allow access to vault credentials" would be some actions applicable to my examples, that would otherwise require a 2 step prompt.

**Reply by Mikenux** [reply]
The app name is required because the user needs to know which app is making the request. This is not a reason. I'm not a developer, but you can check the portal documentation to learn more about portals.

I don't know what a false reason might be, especially since in this case it seems that the app cannot obtain personal data. This is just to stay consistent for portals.

> "use ssh key"/"approve device login"/"allow access to vault credentials"

When are these triggered? On app startup (app unlock)? Via a button (but the button should already indicate which action)?

**Reply by quexten** [reply]
> When are these triggered? On app startup (app unlock)? Via a button (but the button should already indicate which action)?

Well for my use-cases, there is a regular unlock on app startup. The rest don't require the cryptographic key, just authentication, f.e "use-ssh-key" is triggered when the user executes some app that talks to the ssh-agent (ssh, rsync, git sign), in order to authenticate access to the ssh keys.

**Reply by Mikenux** [reply]
For unlocking on app startup, the portal UI can say "Unlock App_Name" if we have the functioning behind it, e.g. kill the app if the identity is not verified (maybe have a given number of tries before killing it).

For SSH:
- There is the request for a portal: https://github.com/flatpak/xdg-desktop-portal/issues/284
- If the ssh-agent is a system component and the desktop environment knows, for example, how to create or give app-specific keys, then it is possible to have actions (via a portal) to request creation and retrieval of app-specific keys. So a portal UI will be able to say "create/generate SSH keys" and "retrieve SSH keys".

You need to know if the action you want is possible on the system side and make it an option/method of a portal, keeping in mind that a portal must ideally only give to the app the specific object requested.

**Reply by quexten** [reply]
For my use-case there is no need for actual permissions to external systems (ssh daemon) to be granted. It's just authentication of the present user, to the application. (In my case the application itself is another SSH daemon, but does not interact with the system's ssh daemon).

So it's really just about verifying that the correct user is present (with password/biometrics), and optionally providing a system secret. To inform the user what this verification is for, a message (such as "access ssh keys") is nice, but this does not relate to any external permissions (access to the system ssh daemon), only to what the app wants to do internally.

**Reply by Mikenux** [reply]
If it's internal to the app, then no, because portals are just an interface for an app to access objects outside of its sandbox.

If the user triggers an internal action, then it is with a button that logically tells them which action will be triggered (for example a "Use SSH" button). Are you worried that the user will forget the context?

**Reply by quexten** [reply]
> Are you worried that the user will forget the context?

I have some cases where the context is not necessarily obvious. For instance, if the user uses ssh (or uses it involuntarily by having an ssh mount, using rsync, opening vs-code/intellij in remote mode, git signing), the action would not be obvious. So just an "authenticate" prompt popping up would be confusing, so having an authenticate prompt together with a message for what action it is would be nice.

This is also solvable of course with the application first showing a "Do you want to authorize XYZ [cancel] [authorize]" prompt, and then on hitting authorize, it would just opening the "application xyz wants to authenticate you" prompt. In that 2-dialog style no message would be required.

I guess for a first version of this portal it is easier to just have an authentication portal that authenticates you, without messages.

**Reply by Mikenux** [reply]
It is possible to have an app-reported reason if you find wording that clearly communicates to users that the portal is not the culprit in case of app misbehavior (e.g. the reported action is not performed).

**Reply by bb010g** [reply]
Would the following sort of dialog work?

***

\[Title: Authentication required\]\
“\<Application\>” is requesting to verify the identity “\<identity\>”.

“\<Application\>” claims this is to “\<reason\>”.

\[Button: Cancel\] \[Button: Authenticate\]

***

If the application does not provide a reason, that respective paragraph could be skipped. If the application does provide a reason, the claimed reason is clearly from the application and is clearly separate from information about the identity.

**Reply by Mikenux** [reply]
Of course, having something like this seems like the way to go.

Additionally (I'm working on other mockups) and if I'm not mistaken, I would clarify in the dialog that the identity is not given to the app, because this doesn't seem to be the app that verifies the identity but the system. Additionally, even if credentials are given to the app, they are credentials that appear to have been created for the app itself, meaning they can be given to it - and this can also be mentioned in this case.

**Reply by bb010g** [reply]
> Additionally (I'm working on other mockups) and if I'm not mistaken, I would clarify in the dialog that the identity is not given to the app, because this doesn't seem to be the app that verifies the identity but the system.

I should note that I pulled that wording from @snwh's <https://gitlab.gnome.org/Teams/Design/os-mockups/-/issues/232#note_1874679>.

**Reply by Mikenux** [reply]
Okay. Feel free to help on this as well. It's helpful to have other wording suggestions to better reflect what is being requested by apps.
